File: distinctSubs.py
class Solution:
    def numDistinct(self, s: str, t: str) -> int:
        n, m = len(s), len(t)

        dp = [1] + [0] * m

        for ch in s:
            for j in range(m-1, -1, -1):
                if ch == t[j]:
                    dp[j+1] += dp[j]
        return dp[m]

File: distinctSubs.rs
impl Solution {
    pub fn num_distinct(s: String, t: String) -> i32 {
        let n = s.len();
        let m = t.len();

        let mut dp = vec![0; m+1];
        dp[0] = 1;

        for &ch in s.as_bytes(){
            for i in (0..m).rev(){
                if ch == t.as_bytes()[i]{
                    dp[i+1] += dp[i];
                }
            }
        }
        return dp[m]
        
    }
}
